src/main/java/net/snowflake/ingest/streaming/internal/DataValidationUtil.java

```diff
@@ -117,72 +138,51 @@ static String validateAndParseObject(Object input) {
-    try {
-      String valueString = getStringValue(input);
-
-      long epoch;
-      int fraction;
-
-      SFTimestamp timestamp = snowflakeDateTimeFormatter.parse(valueString);
-      // The formatter will not correctly parse numbers with decimal values
-      if (timestamp != null) {
-        epoch = timestamp.getSeconds().longValue();
-        fraction = getFractionFromTimestamp(timestamp);
-      } else {
-        /*
-         * Assume the data is of the form "<epoch_seconds>.<fraction of second>". For example "10.5"
-         * is interpreted as meaning 10.5 seconds past the epoch. The following logic breaks out and
-         * stores the epoch seconds and fractional seconds values separately
-         */
-        String[] items = valueString.split("\\.");
-        if (items.length != 2) {
-          throw new SFException(
-              ErrorCode.INVALID_ROW,
-              input.toString(),
-              String.format("Unable to parse timestamp from value.  value=%s", valueString));
-        }
-        epoch = Long.parseLong(items[0]);
-        int l = items.length > 1 ? items[1].length() : 0;
-        /* Fraction is in nanoseconds, but Snowflake will error if the fraction gives
-         * accuracy greater than the scale
-         * */
-        fraction = l == 0 ? 0 : Integer.parseInt(items[1]) * (l < 9 ? Power10.intTable[9 - l] : 1);
-      }
-      if (fraction % Power10.intTable[9 - scale] != 0) {
-        throw new SFException(
-            ErrorCode.INVALID_ROW,
-            input.toString(),
-            String.format(
-                "Row specifies accuracy greater than column scale.  fraction=%d, scale=%d",
-                fraction, scale));
-      }
-      return new TimestampWrapper(epoch, fraction, getTimeInScale(valueString, scale));
-    } catch (NumberFormatException e) {
-      throw new SFException(ErrorCode.INVALID_ROW, input.toString(), e.getMessage());
+  static TimestampWrapper validateAndParseTimestampNtzSb16(
+      Object input, int scale, boolean ignoreTimezone) {
+    String valueString;
+    if (input instanceof String) valueString = (String) input;
+    else if (input instanceof LocalDate) valueString = input.toString();
+    else if (input instanceof LocalDateTime) valueString = input.toString();
+    else if (input instanceof ZonedDateTime)
+      valueString =
+          ignoreTimezone
+              ? ((ZonedDateTime) input).toLocalDateTime().toString()
+              : DateTimeFormatter.ISO_DATE_TIME.format(((ZonedDateTime) input).toOffsetDateTime());
+    else if (input instanceof OffsetDateTime)
+      valueString =
+          ignoreTimezone
+              ? ((OffsetDateTime) input).toLocalDateTime().toString()
+              : DateTimeFormatter.ISO_DATE_TIME.format((OffsetDateTime) input);
+    else
+      throw new SFException(
+          ErrorCode.INVALID_ROW, input.toString(), "Data type not supported for timestamp");
```

Comment:
Could you add to the comment on the types that will be supported now?

Reply:
I added information about allowed data types both to the exception message, as well to the javadoc comment.